File: python/day05.py
from aoc import *
from itertools import groupby, pairwise
from functools import cmp_to_key


def is_valid_line(line, rules):
    return all(pair in rules for pair in pairwise(line))

def middle_page(line):
    return line[len(line)//2]

def part_1(rules, updates):
    return sum(middle_page(u)
               for u in updates
               if is_valid_line(u, rules))

def part_2(rules, updates):
    rules = set(rules)
    def cmp(a, b):
        return -1 if (a, b) in rules else 1
    return sum(middle_page(sorted(u, key=cmp_to_key(cmp)))
               for u in updates
               if not is_valid_line(u, rules))

def solve(filename):
    data = parse_paragraphs(read_input(filename), ints)
    return part_1(*data), part_2(*data)


solve(5)

File: elixir/day05.exs
defmodule Day05 do
  import Enum

  defp valid_line?(line, rules) do
    line
    |> chunk_every(2, 1, :discard)
    |> all?(&MapSet.member?(rules, &1))
  end

  defp middle_page(line) do
    at(line, div(length(line), 2))
  end

  defp sort_line(line, rules) do
    sort(line, &MapSet.member?(rules, [&1, &2]))
  end

  defp part_1([rules, updates]) do
    updates
    |> filter(&valid_line?(&1, rules))
    |> map(&middle_page/1)
    |> sum
  end

  defp part_2([rules, updates]) do
    updates
    |> reject(&valid_line?(&1, rules))
    |> map(&sort_line(&1, rules))
    |> map(&middle_page/1)
    |> sum
  end

  def solve(filename) do
    import AOC

    filename
    |> read_input
    |> parse_paragraphs(:ints)
    |> map(&MapSet.new/1)
    |> juxt([&part_1/1, &part_2/1])
  end
end

Day05.solve(5)
